libass/ass_render.c

```diff
@@ -1185,93 +1347,56 @@ transform_3d(ASS_Vector shift, ASS_Outline *outline, int n_outlines,
  * Tries to get glyph bitmaps from bitmap cache.
  * If they can't be found, they are generated by rotating and rendering the glyph.
  * After that, bitmaps are added to the cache.
- * They are returned in info->bm (glyph), info->bm_o (outline) and info->bm_s (shadow).
+ * They are returned in info->image (glyph), info->image_o (outline).
  */
 static void
 get_bitmap_glyph(ASS_Renderer *render_priv, GlyphInfo *info)
 {
-    if (!info->outline || info->symbol == '\n' || info->symbol == 0 || info->skip)
+    if (!info->outline || info->symbol == '\n' || info->symbol == 0 || info->skip) {
+        ass_cache_dec_ref(info->outline);
+        ass_cache_dec_ref(info->border);
         return;
+    }
 
-    BitmapHashValue *val = ass_cache_get(render_priv->cache.bitmap_cache, &info->hash_key, render_priv);
-    if (!val || !val->valid)
-        info->symbol = 0;
-    info->image = val;
+    double m[3][3];
+    calc_transform_matrix(render_priv, info, m);
+
+    info->image   = get_bitmap(render_priv, info->outline, &info->outline_transform, m);
+    info->image_o = get_bitmap(render_priv, info->border,  &info->border_transform,  m);
 }
 
 size_t ass_bitmap_construct(void *key, void *value, void *priv)
 {
     ASS_Renderer *render_priv = priv;
-    BitmapHashKey *bitmap_key = key;
+    BitmapHashKey *k = key;
     BitmapHashValue *v = value;
-    v->bm = v->bm_o = NULL;
-
-    const size_t hdr = sizeof(BitmapHashKey) + sizeof(BitmapHashValue);
+    v->bm = NULL;
 
-    if (bitmap_key->type == BITMAP_CLIP) {
-        ASS_Rect cbox;
-        ASS_Outline outline;
-        if (!ass_drawing_parse(&outline, &cbox, bitmap_key->u.clip.text, render_priv->library)) {
-            ass_msg(render_priv->library, MSGL_WARN,
-                    "Clip vector parsing failed. Skipping.");
-            return hdr;
-        }
+    double m[3][3];
+    restore_transform(m, k);
 
-        ASS_DVector scale;
-        double w = render_priv->font_scale / (1 << (bitmap_key->u.clip.scale - 1));
-        scale.x = render_priv->font_scale_x * w;
-        scale.y = w;
+    ASS_Outline outline[2];
+    outline_copy(&outline[0], &k->outline->outline[0]);
+    outline_copy(&outline[1], &k->outline->outline[1]);
 
-        ASS_DVector offset;
-        offset.x = int_to_d6(render_priv->settings.left_margin);
-        offset.y = int_to_d6(render_priv->settings.top_margin);
+    for (int i = 0; i < 2; i++) {
+        ASS_Vector *p = outline[i].points;
+        for (size_t j = 0; j < outline[i].n_points; ++j) {
+            double v[3];
+            for (int k = 0; k < 3; k++)
+                v[k] = m[k][0] * p[j].x + m[k][1] * p[j].y + m[k][2];
 
-        ASS_Outline *ol = &outline;
-        for (size_t i = 0; i < ol->n_points; i++) {
-            ol->points[i].x = lrint(ol->points[i].x * scale.x + offset.x);
-            ol->points[i].y = lrint(ol->points[i].y * scale.y + offset.y);
+            double w = 1 / FFMAX(v[2], 0.1);
```

Comment:
I've considered to replicate VSFilter behavior but decided against it. It's due to the following reasons:
1) Old code wasn't scale-invariant. Even if it can replicate VSFilter misfeature at native resolution, it looks significantly different for different output resolutions.
2) It's not some rendering feature, it's mostly for safety: to prevent division by zero and too large images. It's quite hard to trigger in real world scenarios and even if some animation triggers it accidentally it mostly goes unnoticed as problem points would be off screen or fast moving due to large scale factor.
3) To replicate that misfeature it's necessary to add additional degree of freedom in cache key, and that hurts cacheability of regular cases. So some complex logic to distinguish regular and border cases will be necessary.

Therefore I decided to threat is as safety feature and implement similar safety feature of my own. Should I add that whole description to commit message?

Reply:
1. That’s why I said we should multiply the 1000 by `blur_scale`. XySubFilter seems to do this properly, based on my limited testing.

2. You know I’m very conservative when it comes to VSFilter compatibility, and you _may_ be right that this is not needed. I certainly agree that it’s probably not an intentional rendering feature, but I’m not sure it’s quite so hard to trigger. Even if the clipped control point is off-screen, it can still affect what we do see on screen. In the example above, I did intentionally put all points on screen to make the problem extra obvious, but even if you change the `\fry` such that the 3D object extends far beyond the screen frame, the dark and light green areas stay misaligned.

3. To keep the current behaviour, I would just keep `bitmap_advance` as part of the cache key as it is now. It hurts cacheability, yes, but it’s unclear to me just how much. To be honest, I could say the same about the rest of the consolidation, too: do you have examples where it actually helps?

Reply:
This _may_ also be a case of “libass is (well, will be) actually behaving better than VSFilter and nobody is actually using this _intentionally_ in their scripts and _everybody_ would be happy if libass kept behaving better and even VSFilter was updated to match libass rather than vice versa”, like stroking before vs after transforms & rasterization. I don’t know.

Reply:
It's not that hard to trigger intentionally, but I doubt that there are subs where Z-clipping is visible for significant time period. It's possible to calculate correct clipping value and cache it but I consider that it's unnecessary complication. Also if we ever implement GPU rendering the logic of Z-clipping can be significantly different (i. e. nearest parts isn't distorted but removed).

Reply:
As for examples, I started this PR after encountering slowly creeping text with heavy outline generation. While text was barely moving stroking (and also blur) is triggered every frame. Specific speedup depends on output resolution and movement speed, but it can be up to several times. I think I can construct cases with great gain due to any quantization step in this commit.